### PR TITLE
Improve `IncludeExcludeSet` testing

### PR DESCRIPTION
**Pull request by joakime**
Improve testing and code coverage of `IncludeExcludeSet`.

Updating `IncludeExcludeSet` for best performing `.test(input)` method.
(jmh testing of different algorithms shows this one to be fastest)

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java [hunk 1]
```diff
@@ -154,25 +157,47 @@ public void exclude(T... element)
...
+     * <p>
+     *     Empty includes means all inputs are allowed.
+     * </p>
+     *
+     * @param t the input argument
+     * @return true if the input matches an include, and is not excluded.
+     */
     @Override
-    public boolean test(P t)
+    public boolean test(P t) // TestNew
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: removed `// TestNew` comment

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java [hunk 2]
```diff
@@ -154,25 +157,47 @@ public void exclude(T... element)
...
     {
-        if (!_includes.isEmpty() && !_includePredicate.test(t))
-            return false;
+        if (!_includes.isEmpty())
+        {
+            // We have includes defined, if input has no match, return false immediately
+            if (!_includePredicate.test(t))
+                return false;
+            // We have an include match, let excludes override if needed.
+        }
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: I don't find the cascading ifs with hanging comment clearer.  How about:
```suggestion
        // If we have defined includes, but none match then
        if (!_includes.isEmpty() && !_includePredicate.test(t))
        {
            // return false immediately, no need to check excluded
            return false;
        }
        
        // We are included by omission or  match, so we now check excludes
```

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java (hunk 2 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: This is actually one of the implementations we tried.

This was the "simplified" algorithm from ...
https://github.com/eclipse/jetty.project/blob/fix/jetty-12.0.x/improve-includeexcludeset/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java#L160-L164

It (surprisingly) is 10% slower in the jmh testing than what we are using now.
I'll be happy to adjust the comments though.

Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: It _should be_ identical, but it wasn't.

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java (hunk 1 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: Removed.

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java (hunk 2 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: I'm very dubious of the test if they are not identical!

**Review by gregw — APPROVED** [review]

File: jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java (hunk 2 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java]: > I'm very dubious of the test if they are not identical!

You aren't the only one, even @sbordet and @lorban had the same initial reaction, and even with some jmh test tweaking couldn't make it identical.
Even testing both techniques side by side on the same jmh run showed the difference! (consistently the same difference each jmh run)

**Pull request merged**